python/tvm/relax/block_builder.py

```diff
@@ -196,7 +196,9 @@ def _convert_te_arg(self, te_args: Any) -> typing.Tuple[Any, List[tvm.te.Tensor]
         te_args_list = []
 
         def _convert_te_arg_helper(arg):
-            if isinstance(arg, Expr):
+            if isinstance(arg, ShapeExpr):
+                return [_convert_te_arg_helper(x) for x in arg.values]
```

Comment:
This change is a hack to make argument converter support converting a ShapeExpr, which appears as a case when importing UNet from PyTorch FX to high-level operators to CallTIRs.

### Example

Say the API of the high-level operator `reshape` is https://github.com/mlc-ai/relax/blob/05914dbc186f96b9966e9450acfcc2a53022b327/python/tvm/relax/op/transform.py#L52-L54
where `newshape` being a ShapeExpr a is a legal case. However, as one can see, the argument converter will treat the input ShapeExpr as a variable of tensor type and thereby attempt to create a TE tensor for this ShapeExpr, which is not right.

Therefore, here I hack to tackle ShapeExpr in a special way.


### Underlying problem

However, this hack doesn’t solve problems permanently, because the underlying issue is that the converter assumes that **all arguments of type Expr are of type DynTensorType**, which is revealed by the fact that the converter always tries to convert an Expr argument to a TE tensor.

This behavior cannot be correct, since an Expr instance can have various possibilities. It can be
* a Var of type DynTensorType,
* a Tuple of Vars of type DynTensorType,
* a Var which has type TupleType and whose field types are all DynTensorTypes,
* a ShapeExpr, whose values are all PrimExpr,
* a Var which has type ShapeType,
* ...

The status quo is that currently the converter only supports the first possibility.

I once tried to support the second case (“a Tuple of Vars of type DynTensorType”) nearly one month ago, by making the function `te_tensor` able to take a Tuple of Vars and return an array of TE tensors, one for each Var in the input Tuple. I failed on some issue of setting names and values to the returning RXPlaceholderOp (see the code snippet below) and gave up since there are other alternative quick hack. I don’t think I took this problem seriously at that time, and at this moment, I still think enhancing the function `te_tensor` is possible. We need to think of this problem thoroughly.

https://github.com/mlc-ai/relax/blob/05914dbc186f96b9966e9450acfcc2a53022b327/src/relax/ir/emit_te.cc#L39-L42

Apart from supporting DynTensorType in various cases, we should also pay attention to ShapeExpr and other cases.

---

cc @YuchenJin

Reply:
@MasterJH5574 @YuchenJin This should be an updated patch to emit-te so we can recursively convert not only Expr with DynTensor type, but also Tuple, and ShapeExpr.